libevm/VM.cpp

```diff
@@ -374,72 +416,106 @@ void VM::interpretCases()
         }
         NEXT
 
         CASE(LOG1)
         {
             ON_OP();
-            if (m_ext->staticCall)
+            if (m_message->flags & EVM_STATIC)
                 throwDisallowedStateChange();
 
             logGasMem();
             updateIOGas();
 
-            m_ext->log({m_SP[2]}, bytesConstRef(m_mem.data() + (uint64_t)m_SP[0], (uint64_t)m_SP[1]));
+            uint8_t const* data = m_mem.data() + size_t(m_SP[0]);
+            size_t dataSize = size_t(m_SP[1]);
+
+            evm_uint256be topics[] = {toEvmC(m_SP[2])};
+            size_t numTopics = sizeof(topics) / sizeof(topics[0]);
+
+            m_context->fn_table->emit_log(
+                m_context, &m_message->destination, data, dataSize, topics, numTopics);
         }
         NEXT
 
         CASE(LOG2)
         {
             ON_OP();
-            if (m_ext->staticCall)
+            if (m_message->flags & EVM_STATIC)
                 throwDisallowedStateChange();
 
             logGasMem();
             updateIOGas();
 
-            m_ext->log({m_SP[2], m_SP[3]}, bytesConstRef(m_mem.data() + (uint64_t)m_SP[0], (uint64_t)m_SP[1]));
+            uint8_t const* data = m_mem.data() + size_t(m_SP[0]);
+            size_t dataSize = size_t(m_SP[1]);
+
+            evm_uint256be topics[] = {toEvmC(m_SP[2]), toEvmC(m_SP[3])};
+            size_t numTopics = sizeof(topics) / sizeof(topics[0]);
+
+            m_context->fn_table->emit_log(
+                m_context, &m_message->destination, data, dataSize, topics, numTopics);
         }
         NEXT
 
         CASE(LOG3)
         {
             ON_OP();
-            if (m_ext->staticCall)
+            if (m_message->flags & EVM_STATIC)
                 throwDisallowedStateChange();
 
             logGasMem();
             updateIOGas();
 
-            m_ext->log({m_SP[2], m_SP[3], m_SP[4]}, bytesConstRef(m_mem.data() + (uint64_t)m_SP[0], (uint64_t)m_SP[1]));
+            uint8_t const* data = m_mem.data() + size_t(m_SP[0]);
+            size_t dataSize = size_t(m_SP[1]);
+
+            evm_uint256be topics[] = {toEvmC(m_SP[2]), toEvmC(m_SP[3]), toEvmC(m_SP[4])};
+            size_t numTopics = sizeof(topics) / sizeof(topics[0]);
+
+            m_context->fn_table->emit_log(
+                m_context, &m_message->destination, data, dataSize, topics, numTopics);
         }
         NEXT
 
         CASE(LOG4)
         {
             ON_OP();
-            if (m_ext->staticCall)
+            if (m_message->flags & EVM_STATIC)
                 throwDisallowedStateChange();
 
             logGasMem();
             updateIOGas();
 
-            m_ext->log({m_SP[2], m_SP[3], m_SP[4], m_SP[5]}, bytesConstRef(m_mem.data() + (uint64_t)m_SP[0], (uint64_t)m_SP[1]));
+            uint8_t const* data = m_mem.data() + size_t(m_SP[0]);
+            size_t dataSize = size_t(m_SP[1]);
+
+            evm_uint256be topics[] = {
+                toEvmC(m_SP[2]), toEvmC(m_SP[3]), toEvmC(m_SP[4]), toEvmC(m_SP[5])};
```

Comment:
I wonder if the `LOG`s could be simplified now into a single switch case?

Reply:
Probably yes, but I don't want to spend time on this VM implementation more than to make it work via EVM-C. A lot of cleanups and optimization can be done here, but I'd rather implement it from scratch with a different license.